pkg/server/endpoints/node/handler.go

```diff
@@ -369,6 +375,44 @@ func (h *Handler) isAttested(ctx context.Context, baseSpiffeID string) (bool, er
 	return false, nil
 }
 
+func (h *Handler) validateAgentSVID(ctx context.Context, cert *x509.Certificate) error {
+	dataStore := h.c.Catalog.DataStores()[0]
+
+	agentID, err := getSpiffeIDFromCert(cert)
+	if err != nil {
+		return err
+	}
+
+	// agent SVIDs must be unexpired and a belong to the attested nodes.
+	// NOTE: gRPC will reuse connections from agents and therefore, we can't
+	// rely on TLS handshakes to verify certificate validity since the
+	// certificate on the connection could have expired after the initial
+	// handshake.
+	if h.hooks.now().After(cert.NotAfter) {
+		return fmt.Errorf("agent %s SVID has expired", agentID)
+	}
+
+	resp, err := dataStore.FetchAttestedNode(ctx, &datastore.FetchAttestedNodeRequest{
+		SpiffeId: agentID,
+	})
+	if err != nil {
+		return err
+	}
+
+	node := resp.Node
+	if node == nil {
+		return fmt.Errorf("agent %s is not attested", agentID)
+	}
+	if node.CertSerialNumber != cert.SerialNumber.String() {
+		return fmt.Errorf("agent %s SVID does not match expected serial number", agentID)
+	}
+	if node.CertNotAfter != cert.NotAfter.Unix() {
```

Comment:
Hmm this one is interesting, I'm not quite sure what kind of abuse we would be preventing here? I don't _think_ it's possible for these two things to get out of sync... but maybe? The primary reason for storing this data on the attested node was so that we could reap old agents

Reply:
I don't mind removing it. Done!